filters/SkewnessBalancingFilter.cpp

```diff
@@ -79,61 +89,32 @@ std::set<PointId> SkewnessBalancingFilter::processGround(PointViewPtr view)
         M1 += delta_n;
         M3 += term1 * delta_n * (n - 2) - 3 * delta_n * M2;
         M2 += term1;
-        skewness[i] = std::sqrt(n) * M3 / std::pow(M2, 1.5);
-    }
-
-    PointId j(0);
-    PointId i(view->size());
-    do
-    {
-        if (skewness[i] <= 0)
+        skewness = std::sqrt(n) * M3 / std::pow(M2, 1.5);
+        if (skewness > 0 && lastSkewness <= 0)
         {
-            j = i;
-            break;
+            setGround(lastPositive, i - 1);
+            lastPositive = i;
         }
-    } while (--i != 0);
-
-    std::set<PointId> groundIdx;
-    for (PointId i = 0; i <= j; ++i)
-        groundIdx.insert(i);
-
-    log()->get(LogLevel::Debug)
-        << "Stopped with " << groundIdx.size()
-        << " ground returns and skewness of " << skewness[j] << std::endl;
-
-    return groundIdx;
+        lastSkewness = skewness;
```

Comment:
Still trying to convince myself that this actually does what we want. It seems entirely possible that as you iterate through the points the skewness could at times be positive and others negative. You allow this no doubt, but it seems as though only those points between a negative skewness and the next positive skewness will be classed ground. The intent of the algorithm is to only lop off those positive skewness values once the whole distribution is known.

Reply:
I believe that this is slightly different implementation of the same algorithm.  I've run the tests on the old and new code and the results match.

What you say is true -- we classify all points between a negative skewness and a positive one as ground.  But we do this repeatedly,  Ultimately this means that we classify all points from the beginning to the last point with non-positive skewness as ground, even if the skewness flips back and forth between non-positive and positive.  I don't think that this is necessary.  I believe that because the points are sorted, skewness will necessarily be monotonically increasing so there will only be one transition to positive, but I wasn't sure when I started so I did this.

Reply:
I'm convinced now. To the question of monotonically increasing skewness, it is not guaranteed, but the logic holds regardless. It was just not immediately obvious to me (overlooked the bookkeeping on the last transition to positive as the start of the range to set as ground).